src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java

```diff
@@ -0,0 +1,432 @@
+        } catch (Exception x) {
+            LOGGER.log(WARNING, "failed to print message to dead " + context, x);
+        }
+    }
+
+    private static void cancel(StepContext context, Integer build) throws IOException, InterruptedException {
+        if (context.isReady()) {
+            println(context, "Canceled since build #" + build + " already got here");
+            Run<?, ?> r = context.get(Run.class);
+            String job = "";
+            if (r != null) { // it should be always non-null at this point, but let's do a defensive check
+                job = r.getParent().getFullName();
+            }
+            throw new FlowInterruptedException(Result.NOT_BUILT, new CanceledCause(job + "#" + build));
+        } else {
+            LOGGER.log(WARNING, "cannot cancel dead #" + build);
+        }
+    }
+
+    private static void cleanUp(Job<?,?> job, String jobName) {
+        Map<Integer, Milestone> milestonesInJob = getMilestonesByOrdinalByJob().get(jobName);
+        assert milestonesInJob != null;
+        Iterator<Entry<Integer, Milestone>> it = milestonesInJob.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<Integer, Milestone> entry = it.next();
+            Set<Integer> inSight = entry.getValue().inSight;
+            Iterator<Integer> it2 = inSight.iterator();
+            while (it2.hasNext()) {
+                Integer number = it2.next();
+                if (job.getBuildByNumber(number) == null) {
+                    // Deleted at some point but did not properly clean up from exit(…).
+                    LOGGER.log(WARNING, "Cleaning up apparently deleted {0}#{1}", new Object[] {jobName, number});
+                    it2.remove();
+                }
+            }
+        }
+        // TODO: remove milestones that no longer exist (because the step or the job was deleted)
+    }
+
+    @SuppressWarnings("unchecked")
+    private static synchronized void load() {
+        if (getMilestonesByOrdinalByJob() == null) {
+            MilestoneStep.DescriptorImpl.milestonesByOrdinalByJob = new TreeMap<String, Map<Integer, Milestone>>();
+            try {
+                XmlFile configFile = getConfigFile();
+                if (configFile.exists()) {
+                    MilestoneStep.DescriptorImpl.milestonesByOrdinalByJob = (Map<String, Map<Integer, Milestone>>) configFile.read();
+                }
+            } catch (IOException x) {
+                LOGGER.log(WARNING, null, x);
+            }
+            LOGGER.log(Level.FINE, "load: {0}", getMilestonesByOrdinalByJob());
+        }
+    }
+
+    private static synchronized void save() {
+        try {
+            getConfigFile().write(getMilestonesByOrdinalByJob());
+        } catch (IOException x) {
+            LOGGER.log(WARNING, null, x);
+        }
+        LOGGER.log(Level.FINE, "save: {0}", getMilestonesByOrdinalByJob());
+    }
+
+    private static XmlFile getConfigFile() throws IOException {
```

Comment:
:ant: If you moved this into `MilestoneStep.DescriptorImpl`, you would not need to reimplement this standard method. That would be the logical place to put the `load` and `save` methods.